pkg/common/ca/ca_test.go

```diff
@@ -0,0 +1,171 @@
+package ca_test
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"testing"
+	"time"
+
+	"github.com/HewlettPackard/galadriel/pkg/common/ca"
+	"github.com/HewlettPackard/galadriel/pkg/common/cryptoutil"
+	"github.com/HewlettPackard/galadriel/test/certtest"
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/jmhodges/clock"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	expectedKeyUsage = x509.KeyUsageKeyEncipherment | x509.KeyUsageKeyAgreement | x509.KeyUsageDigitalSignature
+)
+
+var (
+	expectedExtendedKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth}
+)
+
+func TestNewCA(t *testing.T) {
+	config := newCAConfig(t, clock.NewFake())
+
+	CA, err := ca.New(config)
+	require.NoError(t, err)
+	require.NotNil(t, CA)
+}
+
+func TestSignX509Certificate(t *testing.T) {
+	clk := clock.NewFake()
+
+	config := newCAConfig(t, clk)
+	serverCA, err := ca.New(config)
+	require.NoError(t, err)
+
+	key, err := cryptoutil.CreateRSAKey()
+	require.NoError(t, err)
+	publicKey := key.Public()
+
+	params := ca.X509CertificateParams{
+		PublicKey: publicKey,
+		TTL:       time.Minute,
+		Subject: pkix.Name{
+			Organization: []string{"test-org"},
+			CommonName:   "test-name",
+		},
+	}
+
+	cert, err := serverCA.SignX509Certificate(params)
+	require.NoError(t, err)
+	require.NotNil(t, cert)
+
+	// check the cert was signed by the serverCA
+	err = cert.CheckSignatureFrom(config.RootCert)
+	require.NoError(t, err)
+
+	assert.NotNil(t, cert.SerialNumber)
+	assert.Equal(t, params.Subject.Organization, cert.Subject.Organization)
+	assert.Equal(t, params.Subject.CommonName, cert.Subject.CommonName)
+	assert.Contains(t, cert.DNSNames, params.Subject.CommonName)
+	assert.Equal(t, publicKey, cert.PublicKey)
+	assert.False(t, cert.IsCA)
+	assert.True(t, cert.BasicConstraintsValid)
+	assert.Equal(t, clk.Now().Add(ca.NotBeforeTolerance), cert.NotBefore)
+	assert.Equal(t, clk.Now().Add(params.TTL), cert.NotAfter)
+	assert.Equal(t, cert.KeyUsage, expectedKeyUsage)
+	assert.Equal(t, cert.ExtKeyUsage, expectedExtendedKeyUsage)
+}
+
+func TestSignJWT(t *testing.T) {
+	clk := clock.NewFake()
+	clk.Set(time.Now())
+
+	config := newCAConfig(t, clk)
+	serverCA, err := ca.New(config)
+	require.NoError(t, err)
+
+	params := ca.JWTParams{
+		Issuer:   "test-issuer",
+		Subject:  spiffeid.RequireTrustDomainFromString("test-domain"),
+		Audience: []string{"test-audience-1", "test-audience-2"},
+		TTL:      time.Minute,
+	}
+
+	token, err := serverCA.SignJWT(params)
+	require.NoError(t, err)
+	require.NotNil(t, token)
+
+	claims := &jwt.RegisteredClaims{}
+	parsed, err := jwt.ParseWithClaims(token, claims, func(token *jwt.Token) (interface{}, error) { return serverCA.PublicKey(), nil })
+
```

Comment:
Make it easier to read it.
```suggestion
  getServerPublicKey := func(token *jwt.Token) (interface{}, error) { return serverCA.PublicKey(), nil }
	parsed, err := jwt.ParseWithClaims(token, claims, getServerPublicKey)
	
```